src/control/cmd/daos_agent/procmon.go

```diff
@@ -135,56 +141,115 @@ func NewProcMon(logger logging.Logger) *procMon {
-	p.submitRequest(ctx, Pid, drpc.MethodGetAttachInfo)
+func (p *procMon) RegisterPool(ctx context.Context, Pid int32, poolReq *mgmtpb.PoolMonitorReq) {
+	req := &procMonRequest{
+		pid:            Pid,
+		action:         drpc.MethodPoolConnect,
+		poolUUID:       poolReq.PoolUUID,
+		poolHandleUUID: poolReq.PoolHandleUUID,
+	}
+	p.submitRequest(ctx, req)
 }
 
-func (p *procMon) UnregisterProcess(ctx context.Context, Pid int32) {
-	p.submitRequest(ctx, Pid, drpc.MethodDisconnect)
+func (p *procMon) DisconnectPool(ctx context.Context, Pid int32, poolReq *mgmtpb.PoolMonitorReq) {
+	req := &procMonRequest{
+		pid:            Pid,
+		action:         drpc.MethodPoolDisconnect,
+		poolUUID:       poolReq.PoolUUID,
+		poolHandleUUID: poolReq.PoolHandleUUID,
+	}
+	p.submitRequest(ctx, req)
 }
 
-func (p *procMon) submitRequest(ctx context.Context, Pid int32, Action drpc.MgmtMethod) {
+func (p *procMon) DisconnectProcess(ctx context.Context, Pid int32) {
 	req := &procMonRequest{
 		pid:    Pid,
-		action: Action,
+		action: drpc.MethodDisconnect,
 	}
+	p.submitRequest(ctx, req)
+}
+
+func (p *procMon) submitRequest(ctx context.Context, request *procMonRequest) {
 	select {
 	case <-ctx.Done():
-	case p.request <- req:
+	case p.request <- request:
 	}
 }
 
-func (p *procMon) handleProcessAttach(ctx context.Context, request *procMonRequest) {
+func (p *procMon) handlePoolConnect(ctx context.Context, request *procMonRequest) {
+	p.log.Debugf("Received request to connect pool:%s with handle %s, for pid:%d", request.poolUUID, request.poolHandleUUID, request.pid)
+
 	info, found := p.procs[request.pid]
 
-	if found {
-		p.log.Errorf("Attempted to monitor process %d more than once", request.pid)
+	if !found {
+		child, cancel := context.WithCancel(ctx)
+		info = &procInfo{
+			log:       p.log,
+			pid:       request.pid,
+			cancelCtx: cancel,
+			response:  p.response,
+			handles:   make(map[string]map[string]bool),
+		}
+
+		p.procs[request.pid] = info
+		go info.monitorProcess(child)
+	}
+
+	_, found = info.handles[request.poolUUID]
+	if !found {
+		info.handles[request.poolUUID] = make(map[string]bool)
+	}
+	info.handles[request.poolUUID][request.poolHandleUUID] = true
+
+}
+
+func (p *procMon) handlePoolDisconnect(request *procMonRequest) {
+	p.log.Debugf("Received request to disconnect pool:%s with handle %s, for pid:%d", request.poolUUID, request.poolHandleUUID, request.pid)
+
+	info, found := p.procs[request.pid]
+
+	if !found || len(info.handles) == 0 {
+		p.log.Debugf("Process has no registered pools to disconnect from.")
 		return
 	}
 
-	child, cancel := context.WithCancel(ctx)
-	info = &procInfo{
-		log:       p.log,
-		pid:       request.pid,
-		cancelCtx: cancel,
-		response:  p.response,
+	if info.handles[request.poolUUID][request.poolHandleUUID] {
+		delete(info.handles[request.poolUUID], request.poolHandleUUID)
+		if len(info.handles[request.poolUUID]) == 0 {
+			delete(info.handles, request.poolUUID)
+		}
```

Comment:
It seems like we probably ought to check to see if the process has any more open handles, and if not, remove it from the monitored processes list as well. Otherwise you could get into a situation where we're still monitoring a bunch of processes that are doing something other then interacting with DAOS.

Reply:
I don't think this can be done here. Either we need to pass the cancelation function down into here or maybe check in the process monitoring go routine to see if we have anything in the map and if not then be done.

Reply:
Can't you just call `ps.handleProcessDisconnect()` here?

Reply:
Nevermind figured it out and fixed it.